File: PythonSolutions/Neetcode150/binarysearch_problems/153_find-min-in-rotated-sorted-array.py
from typing import List
class Solution:
    def findMin(self, nums: List[int]) -> int:
        lo, hi = 0, len(nums) - 1
        # Use infinite here to compare with as we don't know the most max number
            # Or can use an arbitrary number in an array
        minNum = float("inf")

        while lo <= hi:
            mid = lo + ((hi - lo) // 2)
            minNum = min(minNum, nums[mid])

            # It is rotated in a way that smaller numbers are on the left of the bigger number in mid
            if nums[mid] > nums[hi]: 
                # So move lo pointer to the right to find smaller numbers
                lo = mid + 1
            # If mid smaller than end, then its rotated in a way that smaller numbers are on the left
                # This could also mean its not even rotated anymore
            else:
                # So move hi pointer to the left to find smaller numbers
                hi = mid - 1
        
        return minNum
        

File: TypescriptSolutions/Neetcode150/binarysearch_problems/153_find-min-in-rotated-sorted-array.ts
const findMin = (nums: number[]): number => {
    let lo = 0;
    let hi = nums.length - 1;
    let minNum = Infinity;

    while (lo <= hi) {
        const mid = lo + Math.floor((hi - lo) / 2);
        minNum = Math.min(minNum, nums[mid]);

        if (nums[mid] > nums[hi]) {
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }

    return minNum;
};